File: clients/dart/cldf/lib/models/route.dart
import 'package:json_annotation/json_annotation.dart';

import 'enums/protection_rating.dart';
import 'enums/route_type.dart';
import 'enums/route_characteristics.dart';
import 'media/media.dart';

part 'route.g.dart';

/// A climbing route or boulder problem
@JsonSerializable(includeIfNull: false)
class Route {
  /// Creates a new [Route] instance
  Route({
    required this.id,
    this.clid,
    required this.locationId,
    this.sectorId,
    required this.name,
    required this.routeType,
    this.routeCharacteristics,
    this.grades,
    this.height,
    this.color,
    this.firstAscent,
    this.protectionRating,
    this.qualityRating,
    this.beta,
    this.gearNotes,
    this.tags,
    this.media,
    this.customFields,
  });

  /// Creates a [Route] from JSON
  factory Route.fromJson(Map<String, dynamic> json) => _$RouteFromJson(json);

  /// Unique identifier
  final int id;

  /// CrushLog ID - globally unique identifier (v1.3.0+)
  final String? clid;

  /// Reference to location
  final int locationId;

  /// Reference to sector
  final int? sectorId;

  /// Route name
  final String name;

  /// Type of route
  final RouteType routeType;

  /// Route protection characteristics (trad, bolted)
  final RouteCharacteristics? routeCharacteristics;

  /// Grades in different systems
  final Map<String, String>? grades;

  /// Height in meters
  final double? height;

  /// Hold color
  final String? color;

  /// First ascent information
  final FirstAscent? firstAscent;

  /// Protection rating
  final ProtectionRating? protectionRating;

  /// Quality rating (0-5)
  final int? qualityRating;

  /// Beta information
  final String? beta;

  /// Specific gear requirements or protection notes
  final String? gearNotes;

  /// Associated tags
  final List<String>? tags;

  /// Media associated with this route (topos, beta videos, etc.)
  final Media? media;

  /// Custom fields
  final Map<String, dynamic>? customFields;

  /// Converts this [Route] to JSON
  Map<String, dynamic> toJson() => _$RouteToJson(this);
}

/// First ascent information
@JsonSerializable(includeIfNull: false)
class FirstAscent {
  /// Creates a new [FirstAscent] instance
  FirstAscent({this.name, this.date, this.info});

  /// Creates a [FirstAscent] from JSON
  factory FirstAscent.fromJson(Map<String, dynamic> json) =>
      _$FirstAscentFromJson(json);

  /// Name of the first ascensionist
  final String? name;

  /// Date of first ascent
  final String? date;

  /// Additional notes
  final String? info;

  /// Converts this [FirstAscent] to JSON
  Map<String, dynamic> toJson() => _$FirstAscentToJson(this);
}

File: clients/java/cldf-java/src/main/java/app/crushlog/cldf/models/Route.java
package app.crushlog.cldf.models;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.List;

import app.crushlog.cldf.models.enums.ProtectionRating;
import app.crushlog.cldf.models.enums.RouteCharacteristics;
import app.crushlog.cldf.models.enums.RouteType;
import app.crushlog.cldf.models.media.Media;
import app.crushlog.cldf.utils.FlexibleDateTimeDeserializer;
import app.crushlog.cldf.utils.FlexibleLocalDateDeserializer;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/** Represents a climbing route or boulder problem. */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Route {

  @JsonProperty(required = true)
  private Integer id;

  /** CrushLog ID - globally unique identifier (v1.3.0+) */
  private String clid;

  @JsonProperty(required = true)
  private Integer locationId;

  private Integer sectorId;

  @JsonProperty(required = true)
  private String name;

  @JsonProperty(required = true)
  private RouteType routeType;

  private RouteCharacteristics routeCharacteristics;

  private Grades grades;

  private Double height;

  private FirstAscent firstAscent;

  private Integer qualityRating;

  private String color;

  private String beta;

  private ProtectionRating protectionRating;

  private String gearNotes;
  private List<String> tags;

  /** Media associated with this route (topos, beta videos, etc.) */
  private Media media;

  /** QR code data for physical route marking (v1.3.0+) */
  private QrCode qrCode;

  @JsonDeserialize(using = FlexibleDateTimeDeserializer.class)
  private OffsetDateTime createdAt;

  @JsonDeserialize(using = FlexibleDateTimeDeserializer.class)
  private OffsetDateTime updatedAt;

  /** Grade information for different grading systems. */
  @Data
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  public static class Grades {
    @JsonProperty("vScale")
    private String vScale;

    private String font;
    private String french;
    private String yds;
    private String uiaa;
  }

  /** First ascent information. */
  @Data
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  public static class FirstAscent {
    private String name;

    @JsonDeserialize(using = FlexibleLocalDateDeserializer.class)
    private LocalDate date;

    private String info;
  }

  /** QR code data for physical route marking (v1.3.0+). */
  @Data
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  public static class QrCode {
    /** Generated QR code data string. */
    private String data;

    /** Public URL for the route. */
    private String url;

    /** IPFS hash for CLDF archive reference. */
    private String ipfsHash;

    /** Blockchain transaction hash for permanent record. */
    private String blockchainTx;

    /** When the QR code was generated. */
    @JsonDeserialize(using = FlexibleDateTimeDeserializer.class)
    private OffsetDateTime generatedAt;
  }
}
